art/attacks/boundary.py

```diff
@@ -63,67 +65,83 @@ def __init__(self, classifier, targeted=False, delta=0.01, epsilon=0.01, step_ad
                   'epsilon': epsilon,
                   'step_adapt': step_adapt,
                   'max_iter': max_iter,
+                  'num_trial': num_trial,
                   'sample_size': sample_size,
-                  'init_size': init_size}
+                  'init_size': init_size,
+                  }
         self.set_params(**params)
 
-    def generate(self, x, y=None):
+    def generate(self, x, y=None, x_adv_init=None):
         """
         Generate adversarial samples and return them in an array.
 
         :param x: An array with the original inputs to be attacked.
         :type x: `np.ndarray`
         :param y: If `self.targeted` is true, then `y` represents the target labels.
         :type y: `np.ndarray`
+        :param x_adv_init: Initial array to act as initial adversarial examples. Same shape as `x`.
+        :type x_adv_init: `np.ndarray`
         :return: An array holding the adversarial examples.
         :rtype: `np.ndarray`
         """
+        # Prediction from the original images
+        preds = np.argmax(self.classifier.predict(x), axis=1)
+
+        # Prediction from the initial adversarial examples if not None
+        if x_adv_init is not None:
+            init_preds = np.argmax(self.classifier.predict(x_adv_init), axis=1)
+        else:
+            init_preds = [None] * len(x)
+            x_adv_init = [None] * len(x)
+
         # Assert that, if attack is targeted, y is provided
         if self.targeted and y is None:
             raise ValueError('Target labels `y` need to be provided for a targeted attack.')
 
         # Some initial setups
-        x_adv = x.astype(NUMPY_DTYPE)
```

Comment:
Would it be possible to use the previous `astype(NUMPY_TYPE)` here instead of `copy()`?
Both methods return a new copy of the array while the `astype` is consistent with all the other attack implementations and makes use if `NUMPY_DTYPE`.